### ISSUE TITLE
question about reverse engineering a windows USB device

### ISSUE THREAD

**Issue by deelerke** [wcooley/usbrevue#2]
I been using your USBreplay tool to control a USB relay board, which only came with a demo app for Windows.
I did the following: 
- using usbpcap for windows, I captured the operation for closing the relay and for openining.
- this gave me two pcap files (which show each 3 packets in Wireshark / Tshark)
- I tried to replay them on my linux box and got the following error:

root@svr:~# cat /root/on.pcap | usbreplay.py --vid 0x16c0 --pid 0x05df
Error occured in replayer run loop. Here's the traceback
Traceback (most recent call last):
  File "/usr/local/bin/usbreplay.py", line 286, in run
    packet = Packet(hdr, pack)
  File "/usr/local/lib/python2.7/dist-packages/usbrevue.py", line 250, in **ini**
    raise RuntimeError("Not a USB Packet")
RuntimeError: Not a USB Packet
[Errno 110] Operation timed out

Any idea what’s wrong? 
Or any forum or irc channel you can recommend me to get help?

More information:

I bought this relay on eBay
http://www.ebay.co.uk/itm/5V-USB-Relay-1-Channel-Programmable-Computer-Control-For-Smart-Home-/271311678366?pt=UK_BOI_Electrical_Components_Supplies_ET&hash=item3f2b6fab9e

When plugged in , the device registers as a HID device:

[    2.096079] usb 3-1: new low-speed USB device number 2 using uhci_hcd
[    2.271117] usb 3-1: New USB device found, idVendor=16c0, idProduct=05df
[    2.271128] usb 3-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[    2.271135] usb 3-1: Product: USBRelay1
[    2.271141] usb 3-1: Manufacturer: www.dcttech.com

I can provide the the two usb captures if someone wants to see them
thanks

**Comment by wcooley** [comment]
Thanks for opening a ticket; I started responding to your e-mail last week but never got back to it.

You are seeing that error because the packet is too short; we expect it to be at least 64 bytes and the packets in your `on.pcap` file are only 28 bytes long. These tools were written for packet streams captured with the "usbmon" interface in the Linux kernel; usbpcap for Windows seems to capture packets in a different format. (The idea was that you would run a Linux host with Windows in a VM with USB pass-through and on the host use tcpdump w/usbmon to capture -- that's what the `usbcap` script automates.)

**Comment by deelerke** [comment]
Thanks W for the reply!!!
I've set the virtualbox up and can now control the relay from within the guest OS. Getting closer....
From the host (Ubuntu), I'm trying the following:

user-OptiPlex-755 usbrevue # lsusb
Bus 009 Device 002: ID 04d9:1133 Holtek Semiconductor, Inc.
Bus 009 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 003: ID 16c0:05df Van Ooijen Technische Informatica HID device except mice, keyboards, and joysticks
Bus 008 Device 002: ID 413c:2003 Dell Computer Corp. Keyboard

so I need usbcap to look on 'Bus 008 Device 003' ....

user-OptiPlex-755 usbrevue # usbcap 8
tcpdump: usbmon8u: No such device exists
(SIOCGIFHWADDR: No such device)
tcpdump failed. Maybe try with sudo?
user-OptiPlex-755 usbrevue #

now... it seems it's looking for a device called "usbmon8u",
but the device is like this:

user-OptiPlex-755 usbrevue # ls /sys/kernel/debug/usb/usbmon
0s  10s  10u  1t  2s  2u  3t  4s  4u  5t  6s  6u  7t  8s  8u  9t
0u  10t  1s   1u  2t  3s  3u  4t  5s  5u  6t  7s  7u  8t  9s  9u

I tried:

user-OptiPlex-755 usbrevue # usbcap /sys/kernel/debug/usb/usbmon/8u
Invalid bus number /sys/kernel/debug/usb/usbmon/8u.
Usage: /usr/local/bin/usbcap [busnum]
user-OptiPlex-755 usbrevue #

what am I missing here ?

thx!

**Comment by wcooley** [comment]
Try running `tcpdump` directly with _usbmon8_ as the device. It's possible that the **u** is no longer required or handled in `tcpdump`:

```
tcpdump -i usbmon8 -U -w out.pcap
```

You can also use `tcpdump -D` to list available interfaces:

```
$ sudo tcpdump -D
1.bluetooth0 (Bluetooth adapter number 0)
2.nflog (Linux netfilter log (NFLOG) interface)
3.nfqueue (Linux netfilter queue (NFQUEUE) interface)
4.em1
5.usbmon1 (USB bus number 1)
6.usbmon2 (USB bus number 2)
7.usbmon3 (USB bus number 3)
8.usbmon4 (USB bus number 4)
9.any (Pseudo-device that captures on all interfaces)
10.lo
```

(You should also note that the `usbcap` script writes the pcap stream to standard output, so you probably don't want it going to your terminal. Should probably add a check for stdout being a tty...)

**Comment by deelerke** [comment]
Yes! got it working now.
I also trimmed the capture a bit with wireshark to remove unnecessary packets.

thanks so much for the help

**Comment by wcooley** [comment]
You're welcome!

**Comment by anyremote** [comment]
2 deelerke:
i've got the same USB relay. Could You please post Your working pcap  files ?

**Comment by deelerke** [comment]
sure .... where do I upload them? here only PNG, JPEG and GIF is allowed

**Comment by anyremote** [comment]
Hmm... If these are text files You can post the content right here.
Or just send directly to my mailbox anyremote (at) mail.ru 
Thanks!

**Comment by deelerke** [comment]
Okay, I send it to your email address

**Comment by snowbiker** [comment]
I'm also trying to get this relay to work in linux. It seems there might be a way to get it working by sending a short string to a hidraw device? Alternately, if the pcap files are short maybe you could just run od -x pcapfile and post it here, to help out anyone else who finds this thread via google, like I did? I've also never used pcap so I'll have to figure out how to use that, but this thread does give me hope! Thanks.

**Comment by deelerke** [comment]
Hello! 

I can send you the pcap files if you like, but GitHub doesn't allow me to attach them, because of the filetype.
I also found another way to control the relay, without using pcap files.
https://github.com/davidbuechi/UsbRelay which works perfect
You might need to edit the source file to match your board.

best regards
Dieter

**Comment by snowbiker** [comment]
Thanks for the pointer, I don't know how I missed that! In the meantime would you mind emailing the pcap files to me? My email address is snowbiker at gmail dot com. Thanks again.

**Comment by snowbiker** [comment]
Thanks. David Buechi's worked perfect once I changed the device from USBRelay2 to USBRelay1 and it's off ready to powercycle my weatherstation when it stops outputting data. Also, I just wanted to point out that using the xxd command which is [for some reason] part of the vim package can repackage data into hex characters. I made these commands for inclusion in this comment using od -t x1 -w112 ~/rel_on.pcap etc:
echo "d4 c3 b2 a1 02 00 04 00 00 00 00 00 00 00 00 00 ff ff 00 00 dc 00 00 00 3a 49 ff 52 f0 2a 02 00 48 00 00 00 48 00 00 00 00 af 88 35 00 88 ff ff 53 02 00 02 04 00 00 00 3a 49 ff 52 00 00 00 00 f0 2a 02 00 8d ff ff ff 08 00 00 00 08 00 00 00 21 09 00 03 00 00 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ff 01 00 00 00 00 00 00" | xxd -r -p >/tmp/rel_on.pcap
echo "d4 c3 b2 a1 02 00 04 00 00 00 00 00 00 00 00 00 ff ff 00 00 dc 00 00 00 44 49 ff 52 c0 82 01 00 48 00 00 00 48 00 00 00 80 d7 dc b3 00 88 ff ff 53 02 00 02 04 00 00 00 44 49 ff 52 00 00 00 00 c0 82 01 00 8d ff ff ff 08 00 00 00 08 00 00 00 21 09 00 03 00 00 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fd 01 00 00 00 00 00 00" |xxd -r -p >/tmp/rel_off.pcap
